### ISSUE TITLE
add non-root user

### ISSUE THREAD

**Issue by allandegnan** [UKHomeOffice/docker-centos-base#5]
I think we should add a non-root user here such as "app" or "daemon" for consumption downstream.  Further, we look at on hardening this image so said user is pretty locked down to their area.  I don't really see an app needs to have permission all over the filesystem?

**Comment by vaijab** [comment]
How would you allow an app to bind to a privileged port?

**Comment by allandegnan** [comment]
I wouldn't.  It doesn't need to.

**Comment by vaijab** [comment]
This is just one example. I'd argue that restricting app behaviour is a good thing.

**Comment by allandegnan** [comment]
So I'm not suggesting we need to mandate use of the user in all circumstances.  If you install nginx and it configures www-data, starts as root and drops down to www-data it isn't a catastrophe if you run it like that.

However in the vast majority of cases I'd rather you just bind port 8080 or similar than bother with using root then dropping privileges.  Access external to the pod would still be over 443, so it would be immaterial to most users what we actually run our services on inside the cluster.

If you're forced to install some binary software that refuses to work without root, then you don't use this user, but otherwise it's there for the developers to use so they don't need to worry about creating a secure user or their app running as root.

Anyways, I'm happy ready to listen to your argument on why adding a restricted user for apps is a bad thing if that's what you're arguing?

**Comment by vaijab** [comment]
> Anyways, I'm happy ready to listen to your argument on why adding a restricted user for apps is a bad thing if that's what you're arguing?

I am absolutely with you on that - to add a user for apps to use, but what I am also suggesting is that we should not restrict apps binding to privileged ports inside a container, but if there is no easy way to do that, then I guess the argument is over - you won :-)

**Comment by allandegnan** [comment]
@vaijab It's not about winning.  I need a user downstream or I'm running a nodejs app as root; bypassing tools whining about this seems like a bad thing.  These are the options:-

1) - Add a reusable user to centos-base
2) - Add a semi-reusable user to nodejs-base
3) - Add a user to my consumption of nodejs-base
4) - Run as root.

@jon-shanks @PurpleBooth @gambol99 @lewismarshall 

Please do tell me of an alternative option otherwise I'm doing a PR for an app user here.

**Comment by jon-shanks** [comment]
I don't think we should pollute the base image with any form of users. I'd expect the downstream container to add the users they need as part of their own dockerfile when it is a requirement. 

We obviously shouldn't be running things as root generally. So that leaves 2 and 3 and to be honest, i am not sure it matters a huge amount whether it is done at 2 or 3

**Comment by jon-shanks** [comment]
alternatively we could assure this with the changes rohith has done on the schema json, so that if the runasuser isn't specified in the JSON to kubernetes (this is separate to the docker container), that runasnotroot would cover

**Comment by allandegnan** [comment]
@jon-shanks: 

So as per previous comment, I think having the user created upstream meaning it's probably always uid 1000 (or something) is a good thing when we're sharing volumes, but lets move it down to 2 or 3.  If I add the user in 2 and set the USER in the docker file, a developer would have to then circumvent that to run as root and thus can resonably be expected to not do that.  

If we leave it to step 3, a self serve user will likely just not know they're sposed to run as a non-privledged user.  Obviously if there's a solution to do that from kubernetes I'm all ears but I still thinking the shared UID is probably a good point. (based on the s3 archive container, mounting in the backup dir, which I'd like to be owned by the same user, more or less.  This may or may not be an issue that's solved by mounting options however)

**Comment by allandegnan** [comment]
So I asked @gambol99 what he thinks and he agrees with @jon-shanks.  I disagree but I'm being outvoted so I'm going to implement the useradd in my 10 or so downstream repos.
